overlord/snapstate/snapstate.go

```diff
@@ -1070,6 +1076,50 @@ func FinishTaskWithRestart(task *state.Task, status state.Status, rt restart.Res
 	return restart.FinishTaskWithRestart(task, status, rt, rebootRequiredSnap, rebootInfo)
 }
 
+var ErrUnexpectedRuntimeFailure = errors.New("unexpected restart at runtime")
+
+// AssertRuntimeFailureRestart asserts whether the current process state
+// indicates a failure at runtime and depending on the current changes state
+// either returns ErrUnexpectedRuntimeFailure to indicate that the failure
+// handling was invoked due to an earlier unexpected process failure at runtime,
+// or nil indicating that snapd should proceed with execution.
+func AssertRuntimeFailureRestart(st *state.State) error {
+	if !isInvokedWithRevert() {
+		return nil
+	}
+	// we were invoked by snap-failure, there could be an ongoing
+	// refresh of the snapd snap which has failed and a revert is
+	// pending, but it could also be the case that the snapd process
+	// just failed at runtime, in which case systemd may have
+	// triggered an on-failure handling
+
+	foundChangeWhichMayTriggerRestart := false
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+
+		// TODO should this verify that link-snap is done, but we're
+		// waiting for auto-connect?
+		var snaps []string
+		if err := chg.Get("snap-names", &snaps); err == nil {
```

Comment:
there might be changes that touch snapd that are not a refresh, in principle we need to check that there's a auto-connect as it's the one calling FinishRestart*